packages/origin-247-certificate/src/offchain-certificate/certificate.aggregate.ts

```diff
@@ -0,0 +1,151 @@
+import {
+    ICertificateEvent,
+    CertificateEventType,
+    CertificateTransferredEvent,
+    CertificateClaimedEvent
+} from './events/Certificate.events';
+import { IIssueCommand, ICertificateReadModel } from '../types';
+import { CertificateErrors } from './errors';
+import { BigNumber } from 'ethers';
+
+export class CertificateAggregate {
+    private certificate: ICertificateReadModel;
+
+    constructor(private readonly events: ICertificateEvent[]) {
+        // events need to be sorted first
+        // field to sort by needs to be added to event
+        events.forEach((event) => this.apply(event));
+    }
+
+    public apply(event: ICertificateEvent): void {
+        switch (event.type) {
+            case CertificateEventType.Issued:
+                this.issue(event);
+                break;
+            case CertificateEventType.Transferred:
+                this.transfer(event as CertificateTransferredEvent);
+                break;
+            case CertificateEventType.Claimed:
+                this.claim(event as CertificateClaimedEvent);
+                break;
+        }
+    }
+
+    private issue(event: ICertificateEvent): void {
+        if (this.isIssued()) {
+            throw new CertificateErrors.CertificateAlreadyIssuedError(
```

Comment:
If you already have dedicated errors, then you can just pass parameters here for these errors, not whole messages

Reply:
And handle actualy message in error